src/OrbitGl/GraphTrack.cpp

```diff
@@ -17,121 +19,151 @@
-  Track::Draw(batcher, text_renderer, current_mouse_time_ns, picking_mode, z_offset);
-  if (values_.empty() || picking_mode != PickingMode::kNone) {
-    return;
+template <size_t Dimension>
+void GraphTrack<Dimension>::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
+  constexpr uint32_t kDepth = 0;
+  std::shared_ptr<TimerChain> timer_chain = timers_[kDepth];
+  if (timer_chain == nullptr) {
+    timer_chain = std::make_shared<TimerChain>();
+    timers_[kDepth] = timer_chain;
   }
 
-  const Color kBlack(0, 0, 0, 255);
-  const Color kWhite(255, 255, 255, 255);
-  float label_z = GlCanvas::kZValueTrackLabel + z_offset;
+  timer_chain->emplace_back(timer_info);
+}
 
-  // Draw label
-  auto previous_point = GetPreviousValueAndTime(current_mouse_time_ns);
-  double value =
-      previous_point.has_value() ? previous_point.value().second : values_.begin()->second;
+template <size_t Dimension>
+std::vector<std::shared_ptr<TimerChain>> GraphTrack<Dimension>::GetAllChains() const {
+  std::vector<std::shared_ptr<TimerChain>> chains;
+  for (const auto& pair : timers_) {
+    chains.push_back(pair.second);
+  }
+  return chains;
+}
 
-  uint64_t first_time = values_.begin()->first;
-  uint64_t label_time = std::max(current_mouse_time_ns, first_time);
-  float point_x = time_graph_->GetWorldFromTick(label_time);
-  double normalized_value = (value - min_) * inv_value_range_;
+template <size_t Dimension>
+float GraphTrack<Dimension>::GetLabelYFromValues(
+    const std::array<double, Dimension>& /*values*/) const {
   float content_height =
-      (size_[1] - layout_->GetTrackTabHeight() - layout_->GetTrackBottomMargin());
-  float point_y = pos_[1] - layout_->GetTrackTabHeight() -
-                  content_height * (1.f - static_cast<float>(normalized_value));
-  std::string text = value_decimal_digits_.has_value()
-                         ? absl::StrFormat("%.*f", value_decimal_digits_.value(), value)
-                         : std::to_string(value);
-  absl::StrAppend(&text, label_unit_);
-  DrawLabel(batcher, text_renderer, Vec2(point_x, point_y), text, kBlack, kWhite, label_z);
+      size_[1] - layout_->GetTrackTabHeight() - layout_->GetTrackBottomMargin() - GetLegendHeight();
+  return pos_[1] - layout_->GetTrackTabHeight() - GetLegendHeight() - content_height / 2.f;
 }
 
-void GraphTrack::DrawSquareDot(Batcher* batcher, Vec2 center, float radius, float z,
-                               const Color& color) {
-  Vec2 position(center[0] - radius, center[1] - radius);
-  Vec2 size(2 * radius, 2 * radius);
-  batcher->AddBox(Box(position, size, z), color);
+template <size_t Dimension>
+std::string GraphTrack<Dimension>::GetLabelTextFromValues(
+    const std::array<double, Dimension>& values) const {
+  const std::array<std::string, Dimension>& series_names = series_.series_names();
+  std::optional<uint8_t> value_decimal_digits = series_.value_decimal_digits();
+  std::string label_unit = series_.value_unit();
```

Comment:
nit: `unit_label` (label for the unit) instead of `label_unit` (unit of the label)

Reply:
Thanks! I changed it as `value_unit` as we also have the `formatted_value` here now.